### PR TITLE
Coopr 199 Ability to suspend jobs

### PR DESCRIPTION
**Pull request by ytolochkevych**

### CODE REVIEW

**Comment by sreevatsanraman** [comment]
No tests for the feature? How do we make sure it works?

File: coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java [hunk 1]
```diff
@@ -368,6 +368,129 @@ public void requestAbortJob(String clusterId, Account account) throws IOExceptio
...
 
   /**
+   * Pause job
+   */
+  public void requestPauseJob(String clusterId, Account account) throws IOException, MissingClusterException {
+    LOG.debug("request to pause job for cluster: {}", clusterId);
+    ClusterStoreView view = clusterStoreService.getView(account);
+    // First read cluster without locking
+    Cluster cluster = getCluster(clusterId, account);
+    LOG.debug("cluster info: {}", String.valueOf(cluster));
```
Comment (albertshau) [inline coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java]: can pass the cluster directly instead of String.valueOf

File: coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java [hunk 2]
```diff
@@ -368,6 +368,129 @@ public void requestAbortJob(String clusterId, Account account) throws IOExceptio
...
+    if (cluster.getStatus() != Cluster.Status.PENDING) {
+      return;
+    }
+
+    // Get latest job
+    ClusterJob clusterJob = clusterStore.getClusterJob(JobId.fromString(cluster.getLatestJobId()));
+    LOG.debug("latest job info: {}", String.valueOf(clusterJob));
```
Comment (albertshau) [inline coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java]: dont need String.valueOf

File: coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java [hunk 3]
```diff
@@ -368,6 +368,129 @@ public void requestAbortJob(String clusterId, Account account) throws IOExceptio
...
+
+    // If job not running, nothing to pause
+    if (clusterJob.getJobStatus() == ClusterJob.Status.FAILED ||
```
Comment (albertshau) [inline coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java]: why is this if block needed?

File: coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java [hunk 4]
```diff
@@ -368,6 +368,129 @@ public void requestAbortJob(String clusterId, Account account) throws IOExceptio
...
+      clusterJob.getJobStatus() == ClusterJob.Status.COMPLETE) {
+      // Reschedule the job.
+      jobQueues.add(account.getTenantId(), new Element(clusterJob.getJobId()));
+      return;
+    }
+
+    // Job can be paused only when CLUSTER_CREATE is RUNNING
+    if (!(clusterJob.getClusterAction() == ClusterAction.CLUSTER_CREATE &&
```
Comment (albertshau) [inline coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java]: we should be able to pause any type of job, not just cluster_create

File: coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java [hunk 5]
```diff
@@ -368,6 +368,129 @@ public void requestAbortJob(String clusterId, Account account) throws IOExceptio
...
+      clusterJob.getJobStatus() == ClusterJob.Status.RUNNING)) {
+      throw new IllegalStateException("Cannot be paused at this time.");
+    }
+
+    ZKInterProcessReentrantLock lock = lockService.getClusterLock(account.getTenantId(), clusterId);
+    lock.acquire();
+    try {
+      cluster = view.getCluster(clusterId);
+      LOG.debug("cluster info: {}", String.valueOf(cluster));
```
Comment (albertshau) [inline coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java]: can omit String.valueOf

Comment (albertshau) [inline coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java]: and in general for all other log messages

File: coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java [hunk 6]
```diff
@@ -368,6 +368,129 @@ public void requestAbortJob(String clusterId, Account account) throws IOExceptio
...
+      if (clusterJob.getJobStatus() == ClusterJob.Status.COMPLETE ||
+        clusterJob.getJobStatus() == ClusterJob.Status.FAILED) {
+        return;
+      }
+
+      clusterJob.setJobStatus(ClusterJob.Status.PAUSED);
+      clusterJob.setStatusMessage("Paused by user.");
+      clusterStore.writeClusterJob(clusterJob);
+      // Reschedule the job.
+      jobQueues.add(account.getTenantId(), new Element(clusterJob.getJobId()));
```
Comment (albertshau) [inline coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java]: don't think we need to do this

File: coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java [hunk 7]
```diff
@@ -368,6 +368,129 @@ public void requestAbortJob(String clusterId, Account account) throws IOExceptio
   }
 
   /**
+   * Pause job
```
Comment (albertshau) [inline coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java]: needs a proper javadoc with parameters and a description of what it does, when something can be paused, etc.

File: coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java [hunk 8]
```diff
@@ -368,6 +368,129 @@ public void requestAbortJob(String clusterId, Account account) throws IOExceptio
...
+      clusterStore.writeClusterJob(clusterJob);
+      // Reschedule the job.
+      jobQueues.add(account.getTenantId(), new Element(clusterJob.getJobId()));
+    } finally {
+      lock.release();
+    }
+  }
+
+  /**
+   * Resume job
```
Comment (albertshau) [inline coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java]: needs a proper javadoc

File: coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java [hunk 9]
```diff
@@ -368,6 +368,129 @@ public void requestAbortJob(String clusterId, Account account) throws IOExceptio
...
+    }
+
+    // Get latest job
+    ClusterJob clusterJob = clusterStore.getClusterJob(JobId.fromString(cluster.getLatestJobId()));
+    LOG.debug("latest job info: {}", String.valueOf(clusterJob));
+
+    // If job is not paused, nothing to resume
+    if (clusterJob.getJobStatus() != ClusterJob.Status.PAUSED) {
+      // Reschedule the job.
+      jobQueues.add(account.getTenantId(), new Element(clusterJob.getJobId()));
```
Comment (albertshau) [inline coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java]: don't think this is required.  Though it won't cause anything to fail, it will cause extra work to be done.

File: coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java [hunk 10]
```diff
@@ -368,6 +368,129 @@ public void requestAbortJob(String clusterId, Account account) throws IOExceptio
...
+      return;
+    }
+
+    // Job can be resumed only when CLUSTER_CREATE
+    if (clusterJob.getClusterAction() != ClusterAction.CLUSTER_CREATE) {
```
Comment (albertshau) [inline coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java]: any type of job should be able to be paused and resumed

File: coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java [hunk 11]
```diff
@@ -368,6 +368,129 @@ public void requestAbortJob(String clusterId, Account account) throws IOExceptio
   }
 
   /**
+   * Pause job
+   */
+  public void requestPauseJob(String clusterId, Account account) throws IOException, MissingClusterException {
+    LOG.debug("request to pause job for cluster: {}", clusterId);
+    ClusterStoreView view = clusterStoreService.getView(account);
+    // First read cluster without locking
+    Cluster cluster = getCluster(clusterId, account);
```
Comment (albertshau) [inline coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java]: it seems like everything that happens before we get the lock is redundant and can be removed

File: coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java [hunk 12]
```diff
@@ -368,6 +368,129 @@ public void requestAbortJob(String clusterId, Account account) throws IOExceptio
...
+    // Job can be paused only when CLUSTER_CREATE is RUNNING
+    if (!(clusterJob.getClusterAction() == ClusterAction.CLUSTER_CREATE &&
+      clusterJob.getJobStatus() == ClusterJob.Status.RUNNING)) {
+      throw new IllegalStateException("Cannot be paused at this time.");
+    }
+
+    ZKInterProcessReentrantLock lock = lockService.getClusterLock(account.getTenantId(), clusterId);
+    lock.acquire();
+    try {
+      cluster = view.getCluster(clusterId);
```
Comment (albertshau) [inline coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java]: use getCluster method to be consistent

File: coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java [hunk 13]
```diff
@@ -368,6 +368,129 @@ public void requestAbortJob(String clusterId, Account account) throws IOExceptio
...
+
+    // Job can be resumed only when CLUSTER_CREATE
+    if (clusterJob.getClusterAction() != ClusterAction.CLUSTER_CREATE) {
+      throw new IllegalStateException("Cannot be resumed at this time.");
+    }
+
+    ZKInterProcessReentrantLock lock = lockService.getClusterLock(account.getTenantId(), clusterId);
+    lock.acquire();
+    try {
+      cluster = view.getCluster(clusterId);
```
Comment (albertshau) [inline coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java]: use the getCluster method to avoid duplicate code with the null cluster check

File: coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java [hunk 14]
```diff
@@ -368,6 +368,129 @@ public void requestAbortJob(String clusterId, Account account) throws IOExceptio
...
+      lock.release();
+    }
+  }
+
+  /**
+   * Resume job
+   */
+  public void requestResumeJob(String clusterId, Account account) throws IOException, MissingClusterException {
+    LOG.debug("request to resume job for cluster: {}", clusterId);
+    ClusterStoreView view = clusterStoreService.getView(account);
```
Comment (albertshau) [inline coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java]: similarly, i think we can remove everything before the lock since it is redundant

File: coopr-ui/client/js/user/describecluster.js [hunk 15]
```diff
@@ -249,6 +249,22 @@ ClusterView.app.controller('ClusterCtrl',
   };
 
   /**
+   * Pause current job.
+   */
+  $scope.pause = function () {
```
Comment (albertshau) [inline coopr-ui/client/js/user/describecluster.js:254]: coopr-ui is getting phased out for coopr-ngui so this work will have to be taken care of there.

**Comment by ytolochkevych** [comment]
@sreevatsanraman  Tests added to to cluster pause feature

File: coopr-ui/client/js/user/describecluster.js (hunk 15 above)
Comment (ytolochkevych) [inline coopr-ui/client/js/user/describecluster.js:254]: how to run coorp with coopr-ngui?

Comment (albertshau) [inline coopr-ui/client/js/user/describecluster.js:254]: i'm also not sure... we can leave this for a separate PR

**Comment by albertshau** [comment]
couple very minor comments, good to merge after

**Comment by ytolochkevych** [comment]
@albertshau fixed. Please merge
